### PR TITLE
Cpp api doc

### PR DESCRIPTION
**Pull request by JanuszL**

### CODE REVIEW

**Comment by JanuszL** [comment]
Depends on https://github.com/NVIDIA/DALI/pull/575

File: docs/advanced_topics.rst [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+In DALI architecture all GPU operators process the whole batch at a time while the CPU operator processes one sample at a time.
+With DALI the memory is only growing, never shrinking. Hence each GPU buffer is as large as the largest possible batch, while the CPU buffers are as large as batch size multiplied by the size of the largest sample. Note that even though the CPU processes one sample at a time per thread, a vector of samples need to reside in the memory.
+This lazy allocation strategy reduces the number of total memory operations, and hence helps in cases when the memory allocation is expensive, like the pinned CPU memory or GPU memory.
+This is most visible for the operators whose output size may differ from sample to sample and from run to run. Operator with the fixed size outputs, such as crop, does not influence the overall memory consumption growth over time
+
+Operator buffer presizing
+-------------------------
+
+The purpose of this functionality is to enable the user to fine-tune the processing pipeline in situations when it is possible to forecast precisely the memory consumption during a DALI run. This results in saving the overhead of some reallocations.
```
Comment (Kh4L) [inline docs/advanced_topics.rst]:
```suggestion
The purpose of this functionality is to enable the user to fine-tune the processing pipeline in situations where it is possible to forecast precisely the memory consumption during a DALI run. This results in saving the overhead of some reallocations.
```

File: docs/advanced_topics.rst [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+DALI pipeline allows buffering one or more batches of data ahead. This becomes important when the data processing time between batches could vary a lot. Default prefetch depth is 2. The user can change this value using the ``prefetch_queue_depth`` pipeline argument. For example, if the variation is bigger then it is recommended to prefetch more ahead.
+
+C++ API
+-------
+
+.. note::
+
+  **This feature is not officially supported and may change without further notice**
+
+DALI offers two levels of C++ API. One that can be used to develop custom operators and one that allows using DALI as a library from the native applications. Both **C++ APIs are considered unstable and experimental and can be changed** without further notice. For the operator development reference please check how existing operators are implemented. For using DALI itself ``PipelineTest`` family of tests could be a good reference.
```
Comment (Kh4L) [inline docs/advanced_topics.rst]: Level of C++ API is quite ambiguous. I don’t know if we can call the custom op API a C++ API. IMO this part should only focus on the  libdali C++ API calls.

Comment (JanuszL) [reply docs/advanced_topics.rst]: Done

File: docs/advanced_topics.rst (hunk 1 above)
Comment (JanuszL) [reply docs/advanced_topics.rst]: It was Raj suggestion and I'm trust his word choosing.

File: docs/advanced_topics.rst [hunk 3]
```diff
@@ -45,3 +45,12 @@ Prefetching queue depth
...
 DALI pipeline allows buffering one or more batches of data ahead. This becomes important when the data processing time between batches could vary a lot. Default prefetch depth is 2. The user can change this value using the ``prefetch_queue_depth`` pipeline argument. For example, if the variation is bigger then it is recommended to prefetch more ahead.
+
+C++ API
+-------
+
+.. note::
+
+  **This feature is not officially supported and may change without advance notice**
+
+The C++ API enables using DALI as a library from native applications. **The API is experimental, unstable and can change without advance notice**. Refer to ``PipelineTest`` family of tests for how to use the C++ API.
```
Comment (mzient) [inline docs/advanced_topics.rst]:
```suggestion
The C++ API enables using DALI as a library from native applications. **The API is experimental, unstable and can change without notice**. Refer to ``PipelineTest`` family of tests for how to use the C++ API.
```

File: docs/advanced_topics.rst [hunk 4]
```diff
@@ -45,3 +45,12 @@ Prefetching queue depth
 
 The purpose of this functionality is to average the processing time between batches when the variation from batch to batch is high.
 DALI pipeline allows buffering one or more batches of data ahead. This becomes important when the data processing time between batches could vary a lot. Default prefetch depth is 2. The user can change this value using the ``prefetch_queue_depth`` pipeline argument. For example, if the variation is bigger then it is recommended to prefetch more ahead.
+
+C++ API
+-------
+
+.. note::
+
+  **This feature is not officially supported and may change without advance notice**
```
Comment (mzient) [inline docs/advanced_topics.rst]:
```suggestion
  **This feature is not officially supported and may change without notice**
```

File: docs/advanced_topics.rst (hunk 3 above)
Comment (JanuszL) [reply docs/advanced_topics.rst]: Done

File: docs/advanced_topics.rst (hunk 4 above)
Comment (JanuszL) [reply docs/advanced_topics.rst]: Done

**Pull request merged**